components/style/selector_impl.rs

```diff
@@ -8,6 +8,73 @@ use selectors::parser::{ParserContext, SelectorImpl};
 pub enum PseudoElement {
     Before,
     After,
+    FirstLine,
+
+    // https://mxr.mozilla.org/mozilla-central/source/layout/style/nsCSSAnonBoxList.h
```

Comment:
Yeah, forgot to mention that.

Would it be fine to have a feature-gated gecko selector implementation inside style?

I say it because I tried doing exactly that, and the code is pretty suboptimal (and way less readable).

For example, this:

```rust
let map = match pseudo_element {
    Some(PseudoElement::Before) => &self.before_map,
    Some(PseudoElement::After) => &self.after_map,
    Some(_) => &self.other_pseudo_elements_map,
    None => &self.element_map,
};
```

Becomes this:

```rust
let map = if let Some(pseudo) = pseudo_element {
    if pseudo == E::Impl::parse_pseudo_element(&ParserContext::new(), "before").unwrap() {
        &self.before_map
    } else if pseudo == E::Impl::parse_pseudo_element(&ParserContext::new(), "after").unwrap() {
        &self.after_map
    } else {
        &self.other_pseudo_elements_map
    }
} else {
    &self.element_map
};
```